qsc/write_vmec_input.py

```diff
@@ -0,0 +1,283 @@
+          return sp(np.mod(phi,2*np.pi/self.nfp))
+       for j_phi in range(N_phi_conversion):
+          # Solve for the phi0 such that r0 + X1 n + Y1 b has the desired phi
+          phi_target = phi_conversion[j_phi]
+          phi0_rootSolve_min = phi_target - 1.0 / self.nfp
+          phi0_rootSolve_max = phi_target + 1.0 / self.nfp
+        #   print('-------------')
+        #   print(Frenet_to_cylindrical_residual_func(phi0_rootSolve_min))
+        #   print(Frenet_to_cylindrical_residual_func(phi0_rootSolve_max))
+        #   print(Frenet_to_cylindrical_residual_func(phi_target))
+          res = root_scalar(Frenet_to_cylindrical_residual_func, bracket=[phi0_rootSolve_min, phi0_rootSolve_max], x0=phi_target)
+        #   print(res)
+        #   print('-------------')
+          ##### SOLVE FOR THE ZEROS OF Frenet_to_cylindrical_residual !!!!!!!
+        #   call quasisymmetry_fzero(Frenet_to_cylindrical_residual, phi0_rootSolve_min, phi0_rootSolve_max, phi_target, &
+        #        rootSolve_relerr, rootSolve_abserr, fzeroFlag)
+          # Note: fzero returns its answer in phi0_rootSolve_min
+          phi0_solution = res.root
+
+          final_R, final_z = Frenet_to_cylindrical_1_point(phi0_solution)
+          R_2D[j_theta,j_phi] = final_R
+          z_2D[j_theta,j_phi] = final_z
+          phi0_2D[j_theta,j_phi] = phi0_solution
+
+    # Fourier transform the result.
+    # This is not a rate-limiting step, so for clarity of code, we don't bother with an FFT.
+    mpol = int(min(N_theta          / 2, mpol1d))
+    ntor = int(min(N_phi_conversion / 2, ntord))
+    mpol_nonzero = mpol
+    RBC = np.zeros((int(2*ntor+1),int(mpol+1)))
+    RBS = np.zeros((int(2*ntor+1),int(mpol+1)))
+    ZBC = np.zeros((int(2*ntor+1),int(mpol+1)))
+    ZBS = np.zeros((int(2*ntor+1),int(mpol+1)))
+    factor = 2 / (N_theta * N_phi_conversion)
+    for j_phi in range(N_phi_conversion):
+        for j_theta in range(N_theta):
+            for m in range(mpol):
+                nmin = -ntor
+                if m==0: nmin = 1
+                for n in range(nmin, ntor):
+                    angle = m * theta[j_theta] - n * self.nfp * phi_conversion[j_phi]
+                    sinangle = np.sin(angle)
+                    cosangle = np.cos(angle)
+                    factor2 = factor
+                    # The next 2 lines ensure inverse Fourier transform(Fourier transform) = identity
+                    if np.mod(N_theta,2) == 0 and m  == (N_theta/2): factor2 = factor2 / 2
+                    if np.mod(N_phi_conversion,2) == 0 and abs(n) == (N_phi_conversion/2): factor2 = factor2 / 2
+                    RBC[n,m] = RBC[n,m] + R_2D[j_theta, j_phi] * cosangle * factor2
+                    RBS[n,m] = RBS[n,m] + R_2D[j_theta, j_phi] * sinangle * factor2
+                    ZBC[n,m] = ZBC[n,m] + z_2D[j_theta, j_phi] * cosangle * factor2
+                    ZBS[n,m] = ZBS[n,m] + z_2D[j_theta, j_phi] * sinangle * factor2
```

Comment:
This indexing is tricky, because python interprets negative n values as values measured from the end, unlike fortran. I think the approach here may work out fine, but I just point this out in case it causes a problem anywhere.